### DISCUSSION TITLE
Big performance impact when inserting null

### DISCUSSION THREAD

**Discussion by Fredrik-M** [sqlalchemy/sqlalchemy#11256] [Usage Questions] [answered]
I was hoping someone can help me understand why I see worse performance when inserting rows with null values vs. the same inserts with non-null values...

I'm using SQLAlchemy 2.0.28 and MS SQL Server 15.0.4312.2.

The following code sample will profile insertion of rows with a float column, passing zero, None and not passing a value.
When running it with n = 1 000 000, I see the following:

**Zero:** `112019537 function calls (112019514 primitive calls) in 78.572 seconds`
**Implicit null:** `99019537 function calls (99019514 primitive calls) in 103.313 seconds`
**None:** `110019537 function calls (110019514 primitive calls) in 109.643 seconds`

### Example code
```
from pathlib import Path
import cProfile as cp
import inspect
import pstats

import sqlalchemy as sqla
import sqlalchemy_utils as sqla_utils
import sqlalchemy.sql.functions as sqlf
import sqlalchemy.orm


host = 'localhost'
database = 'test'
user = 'sa'
password = None


_Base = sqla.orm.declarative_base()


class Record(_Base):

    __tablename__ = 'record'

    id = sqla.Column(sqla.Integer, primary_key=True)

    value = sqla.Column(sqla.Float)


def create_url(
        host: str,
        database: str,
        user: str,
        password: str,
        driver: str = 'ODBC Driver 17 for SQL Server',
        dialect: str = 'mssql+pyodbc',
        **query,
):
    return sqla.engine.URL.create(
        drivername=dialect,
        username=user,
        password=password,
        host=host,
        database=database,
        query={'driver': driver} | query,
    )


def create_engine(database=database, host=host, user=user, password=password):

    if not password:
        password = input(f'Password for {user} at {host}:')

    return sqla.create_engine(
        create_url(host=host, database=database, user=user, password=password),
    )


eng = create_engine()
sess = sqla.orm.Session(eng)


def init():

    if not sqla_utils.database_exists(eng.url):
        sqla_utils.create_database(eng.url)

    for obj in globals().values():
        if obj is not _Base and inspect.isclass(obj) and issubclass(obj, _Base):
            obj.__table__.create(eng, checkfirst=True)


def main(n):

    def _run(name, cls, n, **data):

        sess.commit()

        with cp.Profile() as pr:
            sess.add_all(
                cls(**data)
                for _ in range(n)
            )
            sess.commit()

            with Path(name + '.prof').open('w') as f:
                stats = pstats.Stats(pr, stream=f)
                stats.sort_stats('cumtime')
                stats.print_stats()

    init()

    _run('zero', Record, 10, value=0)
    _run('impl-null', Record, 10)
    _run('expl-null', Record, 10, value=None)

    _run('zero', Record, n, value=0)
    _run('impl-null', Record, n)
    _run('expl-null', Record, n, value=None)
```

**Comment by zzzeek** [comment]
we would assume the means by which the INSERT statement is rendered is the difference, so you would turn on echo=True and look at the SQL being emitted.  I believe for SQL Server, if you INSERT a row that has no values at all, it uses a syntax that looks like "INSERT INTO table DEFAULT VALUES" or something similar, and it has to do that statement individually for each row, that is, 1M statements.  If OTOH your rows have data it should be using [insertmanyvalues](https://docs.sqlalchemy.org/en/20/core/connections.html#insert-many-values-behavior-for-insert-statements).

the cProfile stuff is good here but when [profiling SQLAlchemy](https://docs.sqlalchemy.org/en/20/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application), always look first at SQL statements being emitted and rows being returned; for example 1000 statements vs. 1M will be a great big difference.     cProfile results can show this also if you look at how many calls there are to `cursor.execute()`, `cursor.executemany()` but a simple echo=True shows simple things like this very quickly.

(Assuming that's what's happening here)

**Reply by Fredrik-M** [reply]
Thanks for the reply.

I did actually check with echo=True, guess I should have mentioned that in my original post, sorry. Looks to me like the emitted SQL is identical for all 3 cases (only inserting 2 rows here):

> 2024-04-11 14:51:09,260 INFO sqlalchemy.engine.Engine COMMIT
> 2024-04-11 14:51:09,262 INFO sqlalchemy.engine.Engine BEGIN (implicit)
> 2024-04-11 14:51:09,265 INFO sqlalchemy.engine.Engine INSERT INTO record (value) OUTPUT inserted.id, inserted.id AS id__1 SELECT p0 FROM (VALUES (?, 0), (?, 1)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter
> 2024-04-11 14:51:09,265 INFO sqlalchemy.engine.Engine [generated in 0.00037s (insertmanyvalues) 1/1 (ordered)] (0, 0)
> 2024-04-11 14:51:09,270 INFO sqlalchemy.engine.Engine COMMIT

> 2024-04-11 14:51:09,513 INFO sqlalchemy.engine.Engine BEGIN (implicit)
> 2024-04-11 14:51:09,514 INFO sqlalchemy.engine.Engine INSERT INTO record (value) OUTPUT inserted.id, inserted.id AS id__1 SELECT p0 FROM (VALUES (?, 0), (?, 1)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter
> 2024-04-11 14:51:09,514 INFO sqlalchemy.engine.Engine [cached since 0.2498s ago (insertmanyvalues) 1/1 (ordered)] (None, None)
> 2024-04-11 14:51:09,518 INFO sqlalchemy.engine.Engine COMMIT

> 2024-04-11 14:51:09,536 INFO sqlalchemy.engine.Engine BEGIN (implicit)
> 2024-04-11 14:51:09,537 INFO sqlalchemy.engine.Engine INSERT INTO record (value) OUTPUT inserted.id, inserted.id AS id__1 SELECT p0 FROM (VALUES (?, 0), (?, 1)) AS imp_sen(p0, sen_counter) ORDER BY sen_counter
> 2024-04-11 14:51:09,537 INFO sqlalchemy.engine.Engine [cached since 0.2723s ago (insertmanyvalues) 1/1 (ordered)] (None, None)
> 2024-04-11 14:51:09,539 INFO sqlalchemy.engine.Engine COMMIT

The cProfile reports all show 1000 calls to `execute`, so I guess it's inserting batches of 1000 rows:

> 1000    3.198    0.003    3.198    0.003 {method 'execute' of 'pyodbc.Cursor' objects}

> 1000   32.021    0.032   32.021    0.032 {method 'execute' of 'pyodbc.Cursor' objects}

> 1000   30.042    0.030   30.042    0.030 {method 'execute' of 'pyodbc.Cursor' objects}

But it's taking an order of magnitude longer in the null cases...

Here are the full cProfile outputs:
[zero.prof.txt](https://github.com/sqlalchemy/sqlalchemy/files/14948385/zero.prof.txt)
[impl-null.prof.txt](https://github.com/sqlalchemy/sqlalchemy/files/14948386/impl-null.prof.txt)
[expl-null.prof.txt](https://github.com/sqlalchemy/sqlalchemy/files/14948387/expl-null.prof.txt)

**Reply by zzzeek** [reply]
okey doke, in your profile, this is the driver and/or SQL Server:

zero.prof:

    1000    3.198    0.003    3.198    0.003 {method 'execute' of 'pyodbc.Cursor' objects}


expl-null.prof:

    1000   30.042    0.030   30.042    0.030 {method 'execute' of 'pyodbc.Cursor' objects}

30 sec vs. 3 sec

that makes up the majority of the total difference in time for these two profs, 78 sec vs 109

What then affects these times is what the driver does with the values.  we have some ability to set this using setinputsizes() but it's not clear what would be appropriate here.

cc @gordthompson , pyodbc driver slowness issue

**Reply by Fredrik-M** [reply]
By the way, I'm using UnixODBC 2.3.11 and pyodbc 5.0.1.

I ran my test code with ODBC trace enabled, here are the relevant parts of the output:
[zero.trace.txt](https://github.com/sqlalchemy/sqlalchemy/files/14949132/zero.trace.txt)
[impl-null.trace.txt](https://github.com/sqlalchemy/sqlalchemy/files/14949133/impl-null.trace.txt)
[expl-null.trace.txt](https://github.com/sqlalchemy/sqlalchemy/files/14949134/expl-null.trace.txt)

I think the difference is that the zero insert doesn't produce these calls, which are present in the null cases:

> [ODBC][187688][1712853301.545725][SQLDescribeParam.c][185]
>                 Entry:
>                         Statement = 0x1a5fa80
>                         Parameter Number = 1
>                         SQL Type = 0x7fe34a80aa10
>                         Param Def = 0x7ffd2a64e578
>                         Scale = 0x7ffd2a64e568
>                         Nullable = 0x7ffd2a64e570
> [ODBC][187688][1712853301.546841][SQLDescribeParam.c][338]
>                 Exit:[SQL_SUCCESS]                
>                         SQL Type = 0x7ffd2a64e050                
>                         Param Def = 0x7ffd2a64e140                
>                         Scale = 0x7ffd2a64e230                
>                         Nullable = 0x7ffd2a64e320
> [ODBC][187688][1712853301.546865][SQLDescribeParam.c][185]
>                 Entry:
>                         Statement = 0x1a5fa80
>                         Parameter Number = 2
>                         SQL Type = 0x7fe34a80aa12
>                         Param Def = 0x7ffd2a64e578
>                         Scale = 0x7ffd2a64e568
>                         Nullable = 0x7ffd2a64e570
> [ODBC][187688][1712853301.546878][SQLDescribeParam.c][338]
>                 Exit:[SQL_SUCCESS]                
>                         SQL Type = 0x7ffd2a64e050                
>                         Param Def = 0x7ffd2a64e140                
>                         Scale = 0x7ffd2a64e230                
>                         Nullable = 0x7ffd2a64e320
>

**Reply by gordthompson** [reply, marked as answer]
Yes, that sounds right. Good detective work!

When pyodbc is binding parameters based on the supplied parameter values it can usually deduce the SQL type and/or the C type directly from the Python type. However, if the Python type is `None` then there is no 100% bulletproof SQL/C type to use in all cases, so pyodbc calls `SQLDescribeParam` to see if the driver can help determine the correct SQL/C type to use. If SQLAlchemy is inserting 2 rows using insertmanyvalues

```
INSERT INTO record (value) OUTPUT inserted.id, inserted.id AS id__1
SELECT p0 FROM (VALUES (?, 0), (?, 1)) AS imp_sen(p0, sen_counter)
ORDER BY sen_counter
```

then `SQLDescribeParam` is called twice. Inserting 1000 rows means that `SQLDescribeParam` gets called 1000 times.

Sometimes `create_engine(url, fast_executemany=True)` can speed things up, but it has [its own issues with None values](https://github.com/mkleehammer/pyodbc/issues/741).

**Reply by zzzeek** [reply]
something else you might try, dont use a bound parameter, send SQL null:

```py
from sqlalchemy import null

_run('expl-null', Record, n, value=null())
```

**Reply by Fredrik-M** [reply]
Thank you both very much for your assistance!

Unfortunately, passing `sqlalchemy.null()` causes the engine to emit one insert per row rather than batching them.

Looks like I will either have to come up with some more or less hacky workaround, or just avoid nulls wherever it's acceptable to use a zero/empty value instead...

**Reply by CaselIT** [reply]
have you tried casting the Nones? I'm not sure if it will make any difference, but it may make sqlalchemy use the setinputsize since it has the information about the type

**Reply by Fredrik-M** [reply]
@CaselIT 
I tried passing `use_setinputsizes=True` to the engine, which results in
`{BindParameter('value', symbol('REQUIRED'), type_=Float()): <class 'float'>}`
in `DialectEvents.do_setinputsizes`, however this does not seem to make any difference. 
That is consistent with what's reported in the [thread](https://github.com/mkleehammer/pyodbc/issues/741) that @gordthompson linked.

**Reply by CaselIT** [reply]
Final suggestion, try using a bind parameter setting literal execute to true https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam.params.literal_execute

I think it will be the same as passing `sa.null()` but it's worth a try

**Reply by CaselIT** [reply]
Also for an example on how to manually override setinputsize using the event you can check this that was just posted by mike: 
https://github.com/sqlalchemy/sqlalchemy/discussions/10979#discussioncomment-9095940

**Reply by gordthompson** [reply]
I should add that `fast_executemany=True` will not help in this case. Using just

```python
engine = create_engine(
    connection_url,
    fast_executemany=True,
)
```

will continue to use `insertmanyvalues`, and

```python
engine = create_engine(
    connection_url,
    fast_executemany=True,
    use_insertmanyvalues=False,
)
```

results in individual INSERT statements for each row.

FWIW, I avoid using ORM objects to perform bulk inserts whenever possible.
